src/mapgen_carpathian.cpp

```diff
@@ -0,0 +1,459 @@
+	s16 stone_surface_max_y = generateTerrain();
+
+	// Create heightmap
+	updateHeightmap(node_min, node_max);
+
+	// Init biome generator, place biome-specific nodes, and build biomemap
+	biomegen->calcBiomeNoise(node_min);
+	MgStoneType stone_type = generateBiomes();
+
+	// Generate caverns, tunnels and classic caves
+	if (flags & MG_CAVES) {
+		bool has_cavern = false;
+		// Generate caverns
+		if (spflags & MGCARPATHIAN_CAVERNS)
+			has_cavern = generateCaverns(stone_surface_max_y);
+		// Generate tunnels and classic caves
+		if (has_cavern)
+			// Disable classic caves in this mapchunk by setting
+			// 'large cave depth' to world base. Avoids excessive liquid in
+			// large caverns and floating blobs of overgenerated liquid.
+			generateCaves(stone_surface_max_y, -MAX_MAP_GENERATION_LIMIT);
+		else
+			generateCaves(stone_surface_max_y, large_cave_depth);
+	}
+
+	// Generate dungeons
+	if (flags & MG_DUNGEONS)
+		generateDungeons(stone_surface_max_y, stone_type);
+
+	// Generate the registered decorations
+	if (flags & MG_DECORATIONS)
+		m_emerge->decomgr->placeAllDecos(this, blockseed, node_min, node_max);
+
+	// Generate the registered ores
+	m_emerge->oremgr->placeAllOres(this, blockseed, node_min, node_max);
+
+	// Sprinkle some dust on top after everything else was generated
+	dustTopNodes();
+
+	// Update liquids
+	updateLiquid(&data->transforming_liquid, full_node_min, full_node_max);
+
+	// Calculate lighting
+	if (flags & MG_LIGHT) {
+		calcLighting(node_min - v3s16(0, 1, 0), node_max + v3s16(0, 1, 0),
+				full_node_min, full_node_max);
+	}
+
+	this->generating = false;
+}
+
+
+///////////////////////////////////////////////////////////////////////////////
+
+
+int MapgenCarpathian::getSpawnLevelAtPoint(v2s16 p)
+{
+	s16 level_at_point = terrainLevelAtPoint(p.X, p.Y);
+	if (level_at_point <= water_level || level_at_point > water_level + 32)
+		return MAX_MAP_GENERATION_LIMIT; // Unsuitable spawn point
+	else
+		return level_at_point;
+}
+
+
+float MapgenCarpathian::terrainLevelAtPoint(s16 x, s16 z)
+{
+	float ground = NoisePerlin2D(&noise_base->np, x, z, seed);
+	float height1 = NoisePerlin2D(&noise_height1->np, x, z, seed);
+	float height2 = NoisePerlin2D(&noise_height2->np, x, z, seed);
+	float height3 = NoisePerlin2D(&noise_height3->np, x, z, seed);
+	float height4 = NoisePerlin2D(&noise_height4->np, x, z, seed);
+	float hter = NoisePerlin2D(&noise_hills_terrain->np, x, z, seed);
+	float rter = NoisePerlin2D(&noise_ridge_terrain->np, x, z, seed);
+	float ster = NoisePerlin2D(&noise_step_terrain->np, x, z, seed);
+	float n_hills = NoisePerlin2D(&noise_hills->np, x, z, seed);
+	float n_ridge_mnt = NoisePerlin2D(&noise_ridge_mnt->np, x, z, seed);
+	float n_step_mnt = NoisePerlin2D(&noise_step_mnt->np, x, z, seed);
+
+	int height = -MAX_MAP_GENERATION_LIMIT;
+
+	for (s16 y = 1; y <= 30; y++) {
+		float mnt_var = NoisePerlin3D(&noise_mnt_var->np, x, y, z, seed);
+
+		// Grad at water_level
+		int grad_wl = 1 - water_level;
```

Comment:
Better to use a s32. It can't be a s16 because the difference between 2 s16s can exceed the range of s16.
If you use a 'terrain zero level' parameter 'grad' will often exceed the range of s16.